app/scripts/controllers/detect-tokens.js

```diff
@@ -91,8 +115,14 @@ export default class DetectTokensController {
       await Promise.all(
         tokensSlice.map(async (tokenAddress, index) => {
           const balance = result[index];
-          if (balance && !balance.isZero()) {
-            await this._preferences.addToken(
+          let ignored;
+          if (this.hiddenTokens.length) {
```

Comment:
I don't think we need this `this.hiddenTokens.length` check and `ignored` can be a `const` with the `find` result, then below, do an `ignored.length` check.  If `this.hiddenTokens` is an empty array, the `find` will never be executed, so there's no advantage to the length check here.

Reply:
Happy to hear what @brad-decker or anyone else thinks though.

Reply:
I think you're right, though find just returns one element so no need to check its length, just whether or not its defined.

Reply:
Wondering if this is a more legible approach: 
```
const isIgnored = Boolean(this.hiddenTokens.find((ignoredTokenAddress) =>
            isEqualCaseInsensitive(ignoredTokenAddress, tokenAddress),
          ));
          if (balance && !balance.isZero() && !isIgnored) {
            await this.tokensController.addToken(
              tokenAddress,
              tokenList[tokenAddress].symbol,
              tokenList[tokenAddress].decimals,
            );
          }
```

Reply:
You can't return true here, it would end up returning a boolean value that can't be passed to Promise.all. A safer approach would be to first create a new list that has only NON ignored tokens in it and then do the promise.all on that map.

The reason for that is two fold. First as mentioned above you can't use .map for branching logic here. You need the return values to be all of the same type, and they must be promises. Finally, iterating on a list inside a .map has potential performance issues. You'll be iterating up to n times per item in tokensSlice. Better to reduce the list of what you are iterating on ahead of time. You could use `Set` to quickly get to a list of tokens that are unique and not ignored.

Reply:
I just realized that the checking for ignored was redundant... that check is already done on line 93. Here is what I have now: 
```
const tokensWithBalance = tokensSlice.filter((_, index) => {
        const balance = result[index];
        return balance && !balance.isZero();
      });

      await Promise.all(
        tokensWithBalance.map((tokenAddress) => {
          return this.tokensController.addToken(
            tokenAddress,
            tokenList[tokenAddress].symbol,
            tokenList[tokenAddress].decimals,
          );
        }),
      );
  ```

Reply:
that looks good!

Reply:
✅